execute/table/copy.go

```diff
@@ -16,7 +16,7 @@ import "github.com/influxdata/flux"
 // this happens and prevent it.
 func Copy(t flux.Table) (flux.BufferedTable, error) {
 	if tbl, ok := t.(flux.BufferedTable); ok {
-		return tbl, nil
+		return tbl.Copy(), nil
```

Comment:
This might be a question for @jsternberg.

To mitigate the issue, we can either call `Copy()` as in the diff currently, or we can build a fresh `BufferedTable` by removing this block entirely. The semantic difference that I can see is the former appears to carry the value for `BufferedTable.i` forward, whereas the later would reset it to `0`.

I was not clear on the impact either way, but since `BufferedTable.Copy` is already a thing, I hoped that it accounted for this already.

Reply:
These functions are weird and the naming to them is a bit of a mistake. `execute.CopyTable()` can actually be thought of more as "ensure table is buffered" while `buf.Copy()` is "copy this table so it can be used somewhere else".

The difference is that `execute.CopyTable()` _consumes_ its input. So the test in there so see "is this already a buffered table?" is mostly meant as an optimization to the function's true purpose. `buf.Copy()` will produce a copy of the table so it can be used in multiple places.

The naming isn't the best but this change would cause `execute.CopyTable()` to not work properly in the locations where it is presently used.

Reply:
Based on your description here, perhaps the thing to do is make the `Copy` call closer (hopefully inside) `tableFind`. There's a slight bit of indirection via a function named `objects.NewTable`.

Regardless, I can look for another point in which to make this happen if this is contradictory to the design.

Thoughts?

Reply:
I can add a type assertion to `tableFind` such that when it gets an already buffered table it calls `Copy()` on it before we hand it off to `NewTable()`. This seems sufficient to fix the bug for the identified case, but it'll put me back to square one as far as testing goes :grimacing:

Reply:
I just paired with Jonathan briefly and with his guidance I'm looking at an alternate fix.

Sounds like the implementations for `Do()` and `Done()` are not currently idempotent and they need to be, and it seems like if we truly do need to drain the table in question we need to add a missing `Copy()` (but there is doubt that this is actually needed today).

Reply:
> To mitigate the issue, we can either call Copy() as in the diff currently, or we can build a fresh BufferedTable by removing this block entirely.

After some back and forth, and the discovery of those other idempotent-related issues, the conclusion was to remove the block rather than artificially add a `Copy()` call.

What I didn't realize when I posed the question in the first place is there's a call here which will copy the table already:

https://github.com/influxdata/flux/blob/878a9f1a60837db0e28c473ea0e12a1f730650ed/stdlib/universe/table_fns.go#L251

Adding an extra copy would have been redundant, possibly leading to a leak (though this wasn't clear from the test suite).